jsonschema-core.xml

```diff
@@ -800,49 +800,257 @@
+                        to do so.  If present, the Core vocabulary entry MUST have a value
+                        of true indicating that it is required.  The behavior of a value
+                        of false is undefined.
+                    </t>
+                    <t>
+                        The current URI for the Core vocabulary is:
+                        <eref target="https://json-schema.org/draft-08/vocabularies/core"/>.
+                    </t>
+                    <t>
+                        The current URI for the corresponding meta-schema is:
+                        <eref target="https://json-schema.org/draft-08/core"/>.
+                    </t>
+                </section>
+                <section title="Example Meta-Schema With Vocabulary Declarations">
+                    <figure>
+                        <preamble>
+                            This meta-schema explicitly declares both the Core and Applicator
+                            vocabularies, and combines their meta-schemas with an "allOf".
+                            It additional restricts the usage of the Applicator vocabulary
+                            by forbidding the keyword prefixed with "unevaluated".  It also
+                            describes a keyword, "localKeyord", that is not part of either
+                            vocabulary.  Note that it is its own meta-schema,
+                            as it relies on both the Core vocabulary (as all schemas do)
+                            and the Applicator vocabulary (for "allOf"0.
+                        </preamble>
+                        <artwork>
+<![CDATA[
+{
+  "$schema": "https://json-schema.org/draft-08/core-app-example#",
+  "$id": "https://json-schema.org/draft-08/core-app-example",
+  "$recursiveAnchor": true,
+  "$vocabulary": {
+    "https://json-schema.org/draft-08/vocabularies/core": true,
+    "https://json-schema.org/draft-08/vocabularies/applicator": true
+  }.
+  "allOf": [
+    {"$ref": "https://json-schema.org/draft-08/core"},
+    {"$ref": "https://json-schema.org/draft-08/applicator"}
+  ],
+  "patternProperties": {
+    "^unevaluated.*$": false
+  },
+  "properties": {
+    "$comment": "Not in vocabulary, but validated if used",
+    "localKeyword": {
+      "type": "string"
+    }
+  }
+}
+]]>
+                        </artwork>
+                        <postamble>
+                            As shown above, even though each of the referenced standard
+                            meta-schemas declares its corresponding vocabulary, this new
+                            meta-schema must re-declare them for itself.  It would be
+                            valid to leave the core vocabulary out of the "$vocabulary" keyword,
+                            but it MUST be referenced through the "allOf" keyword in order
+                            for its terms to be validated.  There is no special case for
+                            validation of core keywords.
```

Comment:
If *core* can be left out of `$vocabularies` do we need *applicators*?  Wouldn't that also be inferred through the `allOf`?

Reply:
`$vocabulary` not `$vocabularies` – someone at some point suggested that it should be singular like `type`.  I don't have my heart set on either form, but I decided that was a reasonable analogy, and the set of individual vocabularies listed under `$vocabulary` are combined to form _the_ vocabulary described by the meta-schema, so it seemed to work.

Anyway, on to your actual question :-)

We need _applicators_ because that whole thing was a balancing act between people who believed that applicators are less replaceable and more fundamental than assertions (hence moving them to core) and people who thought they were also totally replaceable (hence making them a separate vocabulary that you must explicitly include).

I'm not re-litigating that.  It was not as traumatic as `unevaluated*` but only because it didn't take multiple years to sort out.

Reply:
@gregsdennis 

> Wouldn't that also be inferred through the allOf?

I tried to work out how this would be implemented once, and the general case is challenging.  You need to determine whether there is a `$ref` to the meta-schema within any dynamic scope chain that starts in the root schema object of the entry point schema, and consists only of in-place applicators (including `$ref` and `$recursiveRef`).  This means that you have to understand which keywords are in-place applicators, including those created by 3rd parties as extensions.

If you have a meta-schema that does a lot of extending and composing, this gets hard to spot for end-users as well.

Reply:
So the `$vocabulary` (see, I did it right) entries are used to validate *this* meta-schema, whereas the `allOf` entries are used to validate schemas that *use* this meta-schema, correct?  I'm not sure the reason why *core* is required (that schemas that use this meta-schema need it) is clear from this paragraph.

Reply:
* A single URI-identified vocabulary reserves some set of keywords, with the semantics defined by whatever specification document defined the vocabulary.  Maybe that's one of our drafts, someone else's W3C recommendation, or some wiki page on your local intranet.  As it stands, the keyword semantics have to be interpreted by a human.
* Putting a URI-identified vocabulary in the `$vocabulary` set of a meta-schema M says that _schemas that declare M as their meta-schema:_ `"$schema": M` are expected to use those keywords according to the vocabulary's semantics
* That meta-schema M must **also** decide what possible schemas should be validated for this combination of vocabularies, and to what degree.
    * It could just compose the meta-schemas published by the vocabulary authors (which should be the minimal reasonable set of validation rules for the vocabulary- the _core.json_ and _applicator.json_ meta-schemas illustrate this.
    * It could be a much more strict set of rules, for example applying `"unevaluatedProperties": false` all over the place to forbid further extension and catch spelling errors.  This is particularly common for test environments, but also some applications aren't interested in evolution or compatibility, they are interested in reliability.
    * It could be a much more lax set of rules- maybe you need to conform to only one or the other vocabulary's meta-schema.  I have no use case for this, but it came up once, I forget why.
  

So, every keyword in a meta-schema **describes the schemas**.  Except `$schema`, which declares the meta-schema's meta-schema.

`$vocabulary`, however, is _not_ and exception.  Just as how `"type": "object", "properties": {...}` applies to the schema (not the meta-schema), `$vocabulary` is a usage declaration ***for the schemas*** not the meta-schema.

I should find some way to emphasize how this is not unusual behavior.  Everything in a schema document affects the document it describes.   In the case of metas-schemas, those documents are schema.  `$schema` is the only exception, identifying a meta-schema that applies to the document containing `$schema`.

Reply:
If `$vocabulary` applies to the schema, not the meta-schema that declares it, what's the difference between that and the `allOf` in the example?  Why have both?

It seems to me that the vocabularies could be inferred from the `allOf` contents.

Conversely, the validation facilitated by the `allOf` could proceed by interpreting the vocabularies declared.

Reply:
I think this is resolved based on our offline conversation.